cli/src/cmd/deploy.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of ink!.
+//
+// ink! is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// ink! is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with ink!.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{
+    cmd::{
+        build,
+        Result,
+    }
+};
+
+use futures::future::Future;
+use runtime_primitives::{
+    generic::{Era, Header},
+    traits::{IdentityLookup, Verify, BlakeTwo256},
+    AnySignature,
+};
+use std::{
+    collections::HashMap,
+    fs,
+    io::Read,
+    path::{Path, PathBuf},
+};
+use substrate_primitives::{
+    crypto::Pair,
+    sr25519,
+    H256,
+};
+use subxt::{
+    balances::Balances,
+    contracts::{
+        Contracts,
+        ContractsXt,
+    },
+    system::System,
+};
+
+/// Load the wasm blob from the specified path, or defaults to the target contract wasm in the
+/// current project, inferred via the crate metadata.
+fn load_contract_code(path: Option<&PathBuf>) -> Result<Vec<u8>> {
+    let default_wasm_path = build::collect_crate_metadata()
+        .map(|metadata| metadata.dest_wasm())?;
```

Comment:
might be replacable by `.map(CrateMetadata::dest_wasm)?`. not sure if that's an improvement though.